Errors after upgrade

Hello,
today I've upgraded jwt-cpp and I have compiler errors:

```
[ 94%] Building CXX object CMakeFiles/nms.dir/src/controller/UserController.cpp.o
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/tokenService.cpp:5:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/userCache.hpp:9,
                 from /home/marcin/projects/nms/back/src/service/userCache.cpp:2:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/userCache.hpp:9,
                 from /home/marcin/projects/nms/back/src/controller/UserController.hpp:7,
                 from /home/marcin/projects/nms/back/src/controller/UserController.cpp:1:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/dto/apiDTO/AuthResponse.hpp:5,
                 from /home/marcin/projects/nms/back/src/dto/apiDTO/AuthResponse.cpp:2:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
make[3]: *** [CMakeFiles/nms.dir/build.make:135: CMakeFiles/nms.dir/src/service/tokenService.cpp.o] Błąd 1
make[3]: *** Oczekiwanie na niezakończone zadania....
make[3]: *** [CMakeFiles/nms.dir/build.make:148: CMakeFiles/nms.dir/src/service/userCache.cpp.o] Błąd 1
make[3]: *** [CMakeFiles/nms.dir/build.make:96: CMakeFiles/nms.dir/src/dto/apiDTO/AuthResponse.cpp.o] Błąd 1
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/userCache.hpp:9,
                 from /home/marcin/projects/nms/back/src/controller/InitController.hpp:12,
                 from /home/marcin/projects/nms/back/src/controller/InitController.cpp:1:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
```

What's wrong?

Hello, when we can expect properly fix it? The problem still occurs :(

```cpp
/home/marcin/projects/nms_new/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
```

I tried to install jwt-cpp (not as submodule) and the same result
```
root@marcin-huawei:/tmp/jwt-cpp/build# cmake ..
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found suitable version "1.1.1g", minimum required is "1.0.2")  
-- Found GTest: /usr/lib/x86_64-linux-gnu/libgtest.a  
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/jwt-cpp/build
root@marcin-huawei:/tmp/jwt-cpp/build# make 
Scanning dependencies of target jwt-cpp-test
[  5%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/BaseTest.cpp.o
[ 11%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/ClaimTest.cpp.o
[ 16%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/Keys.cpp.o
[ 22%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/HelperTest.cpp.o
[ 27%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/TestMain.cpp.o
[ 33%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/TokenFormatTest.cpp.o
[ 38%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/TokenTest.cpp.o
[ 44%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/NlohmannTest.cpp.o
[ 50%] Building CXX object tests/CMakeFiles/jwt-cpp-test.dir/OpenSSLErrorTest.cpp.o
[ 55%] Linking CXX executable jwt-cpp-test
[ 55%] Built target jwt-cpp-test
Scanning dependencies of target rsa-verify
[ 61%] Building CXX object example/CMakeFiles/rsa-verify.dir/rsa-verify.cpp.o
[ 66%] Linking CXX executable rsa-verify
[ 66%] Built target rsa-verify
Scanning dependencies of target rsa-create
[ 72%] Building CXX object example/CMakeFiles/rsa-create.dir/rsa-create.cpp.o
[ 77%] Linking CXX executable rsa-create
[ 77%] Built target rsa-create
Scanning dependencies of target private-claims
[ 83%] Building CXX object example/CMakeFiles/private-claims.dir/private-claims.cpp.o
[ 88%] Linking CXX executable private-claims
[ 88%] Built target private-claims
Scanning dependencies of target print-claims
[ 94%] Building CXX object example/CMakeFiles/print-claims.dir/print-claims.cpp.o
[100%] Linking CXX executable print-claims
[100%] Built target print-claims
root@marcin-huawei:/tmp/jwt-cpp/build# make install
[ 55%] Built target jwt-cpp-test
[ 66%] Built target rsa-verify
[ 77%] Built target rsa-create
[ 88%] Built target private-claims
[100%] Built target print-claims
Install the project...
-- Install configuration: ""
-- Up-to-date: /usr/local/include/jwt-cpp/jwt.h
-- Up-to-date: /usr/local/include/jwt-cpp/base.h
-- Up-to-date: /usr/local/include/picojson/picojson.h
-- Installing: /usr/local/jwt-cpp/jwt-cpp.cmake
-- Installing: /usr/local/jwt-cpp/jwt-cpp-config.cmake
-- Installing: /usr/local/jwt-cpp/jwt-cpp-config-version.cmake
root@marcin-huawei:/tmp/jwt-cpp/build# 
```

And from my project:
```
[ 91%] Building CXX object CMakeFiles/nms.dir/src/service/authService.cpp.o
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/dto/apiDTO/AuthResponse.hpp:5,
                 from /home/marcin/projects/nms/back/src/dto/apiDTO/AuthResponse.cpp:2:
/usr/local/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/usr/local/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/usr/local/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/usr/local/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
```

--- a/include/jwt-cpp/jwt.h
+++ b/include/jwt-cpp/jwt.h
@@ -967,12 +967,28 @@ namespace jwt {
 				size_t len = EVP_PKEY_size(pkey.get());
 				std::string res(len, '\0');
 
+				// LibreSSL is the special kid in the block, as it does not support EVP_DigestSign.
+				// OpenSSL on the otherhand does not support using EVP_DigestSignUpdate for eddsa, which is why we end up with this mess.
+				#ifdef LIBRESSL_VERSION_NUMBER
+				ERR_clear_error();
+				auto e = EVP_DigestSignUpdate(ctx.get(), reinterpret_cast<const unsigned char*>(data.data()), data.size());
+				if(e != 1) {
+					std::cout << ERR_error_string(ERR_get_error(), NULL) << std::endl;
+					ec = error::signature_generation_error::signupdate_failed;
+					return {};
+				}
+				if(EVP_DigestSignFinal(ctx.get(), reinterpret_cast<unsigned char*>(&res[0]), &len) != 1) {
+					ec = error::signature_generation_error::signfinal_failed;
+					return {};
+				}
+				#else
 				if (EVP_DigestSign(ctx.get(),
 						reinterpret_cast<unsigned char*>(&res[0]), &len,
 						reinterpret_cast<const unsigned char*>(data.data()), data.size()) != 1) {
 					ec = error::signature_generation_error::signfinal_failed;
 					return {};
 				}
+				#endif
 
 				res.resize(len);
 				return res;
@@ -995,6 +1011,18 @@ namespace jwt {
 					ec = error::signature_verification_error::verifyinit_failed;
 					return;
 				}
+				// LibreSSL is the special kid in the block, as it does not support EVP_DigestVerify.
+				// OpenSSL on the otherhand does not support using EVP_DigestVerifyUpdate for eddsa, which is why we end up with this mess.
+				#ifdef LIBRESSL_VERSION_NUMBER
+				if (EVP_DigestVerifyUpdate(ctx.get(), reinterpret_cast<const unsigned char*>(data.data()), data.size()) != 1) {
+					ec = error::signature_verification_error::verifyupdate_failed;
+					return;
+				}
+				if (EVP_DigestVerifyFinal(ctx.get(), reinterpret_cast<const unsigned char*>(signature.data()), signature.size()) != 1) {
+					ec = error::signature_verification_error::verifyfinal_failed;
+					return;
+				}
+				#else
 				auto res = EVP_DigestVerify(ctx.get(),
 					reinterpret_cast<const unsigned char*>(signature.data()), signature.size(),
 					reinterpret_cast<const unsigned char*>(data.data()), data.size());
@@ -1002,6 +1030,7 @@ namespace jwt {
 					ec = error::signature_verification_error::verifyfinal_failed;
 					return;
 				}
+				#endif
 			}
 			/**
 			 * Returns the algorithm name provided to the constructor

--- a/include/jwt-cpp/jwt.h
+++ b/include/jwt-cpp/jwt.h
@@ -971,8 +971,7 @@ namespace jwt {
 				// OpenSSL on the otherhand does not support using EVP_DigestSignUpdate for eddsa, which is why we end up with this mess.
 				#ifdef LIBRESSL_VERSION_NUMBER
 				ERR_clear_error();
-				auto e = EVP_DigestSignUpdate(ctx.get(), reinterpret_cast<const unsigned char*>(data.data()), data.size());
-				if(e != 1) {
+				if(EVP_DigestSignUpdate(ctx.get(), reinterpret_cast<const unsigned char*>(data.data()), data.size()) != 1) {
 					std::cout << ERR_error_string(ERR_get_error(), NULL) << std::endl;
 					ec = error::signature_generation_error::signupdate_failed;
 					return {};

